compiler/src/lib.rs

```diff
@@ -71,22 +77,111 @@ pub fn compile(
     write(writer, &bytecode, &cp)
 }
 
-fn compile_instruction_set(
+/// Resolves all captured variables of a given chunk identifier.
+///
+/// This function will resolve all direct captures of the chunk and the captures of its inner chunks.
+/// All resolved captures are set into the given `captures` vector.
+fn resolve_captures(
+    engine: &Engine,
+    relations: &Relations,
+    chunk_id: SourceId,
+    captures: &mut Captures,
+) {
+    if captures[chunk_id.0].is_some() {
+        // if captures are already computed for given chunk then return
+        return;
+    }
+
+    let mut chunks = HashSet::new();
+    let mut externals = IndexSet::new();
+
+    fn resolve(
+        chunk_id: SourceId,
+        engine: &Engine,
+        relations: &Relations,
+        captures: &mut Captures,
+        externals: &mut IndexSet<ResolvedSymbol>,
+        chunks: &mut HashSet<SourceId>,
+    ) {
+        if let Some(captures) = &captures[chunk_id.0] {
+            externals.extend(captures);
+            return;
+        }
+
+        let env = engine.get_environment(chunk_id).unwrap();
+
+        externals.extend(
+            env.variables
+                .external_vars()
+                .map(|(_, relation)| {
+                    relations[relation]
+                        .state
+                        .expect_resolved("unresolved relation during compilation")
+                })
+                .filter(|symbol| {
+                    // keep variables only
+                    let env = engine.get_environment(symbol.source).unwrap();
+                    let var = env.variables.get_var(symbol.object_id).unwrap();
+                    var.ty == TypeInfo::Variable
+                }),
+        );
+
+        for (_, func) in env
+            .variables
+            .iter()
+            .filter(|(_, v)| v.ty == TypeInfo::Function)
+        {
+            let func_fqn = &env.fqn.appended(Name::new(&func.name));
+            let (func_id, _) = engine.find_environment_by_name(func_fqn).unwrap();
+
+            resolve(func_id, engine, relations, captures, externals, chunks)
+        }
+
+        chunks.insert(chunk_id);
+        externals.retain(|symbol| !chunks.contains(&symbol.source));
```

Comment:
Those lines don't seem to make sense. Why would a symbol not be external if already seen? If not, shouldn't the filtering be done only one time at the end?